### PR TITLE
HSEARCH-3841 new version_check.enabled setting added

### PR DESCRIPTION
**Pull request by acucumel**
https://hibernate.atlassian.net/browse/HSEARCH-3841

new settings added to get rid of the version check at start in some cases.

### CODE REVIEW

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchLinkImpl.java [hunk 1]
```diff
@@ -111,12 +114,26 @@ void onStart(ConfigurationPropertySource propertySource) {
...
+					if ( !configuredVersion.matches( elasticsearchVersion ) ) {
+						throw log.unexpectedElasticsearchVersion( configuredVersion, elasticsearchVersion );
+					}
+				}
+			}
+			else if ( configuredVersionOptional.isPresent() ) {
 				ElasticsearchVersion configuredVersion = configuredVersionOptional.get();
-				if ( !configuredVersion.matches( elasticsearchVersion ) ) {
-					throw log.unexpectedElasticsearchVersion( configuredVersion, elasticsearchVersion );
+				if ( configuredVersion.getMinor().isPresent() && configuredVersion.getMicro().isPresent() ) {
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchLinkImpl.java]: Ideally this (and the case "no version is configured") should be checked earlier, in `ElasticsearchBackendFactory`. This is especially important for Quarkus, where the configuration is checked at compile-time.

Also, I think you can relax the constraint and allow the micro to be missing: the Elasticsearch team would have to mess up big time for us to have to behave differently based on the micro.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java [hunk 2]
```diff
@@ -80,8 +81,8 @@
 			value = "Elasticsearch request failed: %3$s\nRequest: %1$s\nResponse: %2$s"
 	)
 	SearchException elasticsearchRequestFailed(
-			@FormatWith( ElasticsearchRequestFormatter.class ) ElasticsearchRequest request,
-			@FormatWith( ElasticsearchResponseFormatter.class ) ElasticsearchResponse response,
+			@FormatWith(ElasticsearchRequestFormatter.class) ElasticsearchRequest request,
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java]: I agree these changes are correct, but would you mind moving them to a separate commit dedicated to formatting changes?

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchLinkImpl.java (hunk 1 above)
Comment (acucumel) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchLinkImpl.java]: thanks for the tips, i will do that in few minutes

**Comment by acucumel** [comment]
changes made

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java (hunk 2 above)
Comment (acucumel) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java]: i tried to remove my first commit but it didn't workout ^^

**Comment by yrodiere** [comment]
Looks you added a test for the case where no version is configured; which is nice, I didn't think about that. But you forgot to add the test where everything works :)

As explained above:

> Copy/paste explicitModelDialect to a new test method that checks what happens when the version check is disabled and a correct version is passed. You should only need to set the new property with a call to withBackendProperty, and to remove the call to elasticsearchClientSpy.expectNext.

Also, the build fails because of style issues: see https://ci.hibernate.org/blue/organizations/jenkins/hibernate-search/detail/PR-2220/2/pipeline#step-38-log-958

You can run `mvn checkstyle:check -fn` locally before you commit to check the code style: that way, you won't have to wait for the CI build.

**Comment by acucumel** [comment]
mvn checkstyle:check -fn is ok this time and i add the `explicitProtocolDialect_noVersionCheck_completeVersion` :)

**Comment by yrodiere** [comment]
I rebased on master, squashed your commits and removed the formatting changes in `Log.java`.

I also changed the "version" documentation (including pre-existing documentation) to focus more on the use case where the version check is disabled; after some though I think it makes more sense.

Merging, thanks!
